# Segmenting and Clustering Neighbourhoods - Toronto

## Section 1

In [1]:
import pandas as pd

In [2]:
import requests

In [3]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
wiki_url = requests.get(url)
wiki_url

<Response [200]>

In [4]:
wiki_data = pd.read_html(wiki_url.text)
wiki_data

[    Postal Code           Borough  \
 0           M1A      Not assigned   
 1           M2A      Not assigned   
 2           M3A        North York   
 3           M4A        North York   
 4           M5A  Downtown Toronto   
 ..          ...               ...   
 175         M5Z      Not assigned   
 176         M6Z      Not assigned   
 177         M7Z      Not assigned   
 178         M8Z         Etobicoke   
 179         M9Z      Not assigned   
 
                                          Neighbourhood  
 0                                         Not assigned  
 1                                         Not assigned  
 2                                            Parkwoods  
 3                                     Victoria Village  
 4                            Regent Park, Harbourfront  
 ..                                                 ...  
 175                                       Not assigned  
 176                                       Not assigned  
 177                

In [5]:
len(wiki_data), type(wiki_data)

(3, list)

In [6]:
wiki_data = wiki_data[0]
wiki_data

Postal Code           Borough  \
0           M1A      Not assigned   
1           M2A      Not assigned   
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
..          ...               ...   
175         M5Z      Not assigned   
176         M6Z      Not assigned   
177         M7Z      Not assigned   
178         M8Z         Etobicoke   
179         M9Z      Not assigned   

                                         Neighbourhood  
0                                         Not assigned  
1                                         Not assigned  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
..                                                 ...  
175                                       Not assigned  
176                                       Not assigned  
177                                       Not assigned  
178  Mimico NW, The Queensway West, South of Bloor,...  
179                                       Not assigned  

[180 rows x 3 columns]

Ignoring 'not assigned' boroughs.

In [7]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [8]:
df = df.groupby(['Postal Code']).head()
df

Postal Code           Borough  \
2           M3A        North York   
3           M4A        North York   
4           M5A  Downtown Toronto   
5           M6A        North York   
6           M7A  Downtown Toronto   
..          ...               ...   
160         M8X         Etobicoke   
165         M4Y  Downtown Toronto   
168         M7Y      East Toronto   
169         M8Y         Etobicoke   
178         M8Z         Etobicoke   

                                         Neighbourhood  
2                                            Parkwoods  
3                                     Victoria Village  
4                            Regent Park, Harbourfront  
5                     Lawrence Manor, Lawrence Heights  
6          Queen's Park, Ontario Provincial Government  
..                                                 ...  
160      The Kingsway, Montgomery Road, Old Mill North  
165                               Church and Wellesley  
168  Business reply mail Processing Centre, South C...  
169  Old Mill South, King's Mill Park, Sunnylea, Hu...  
178  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [9]:
df.Neighbourhood.str.count("Not assigned").sum()

0

Good news - no 'not assigned' neighbourss in dataframe!

In [10]:
df = df.reset_index()
df

index Postal Code           Borough  \
0        2         M3A        North York   
1        3         M4A        North York   
2        4         M5A  Downtown Toronto   
3        5         M6A        North York   
4        6         M7A  Downtown Toronto   
..     ...         ...               ...   
98     160         M8X         Etobicoke   
99     165         M4Y  Downtown Toronto   
100    168         M7Y      East Toronto   
101    169         M8Y         Etobicoke   
102    178         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 4 columns]

In [11]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4          Queen's Park, Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100  Business reply mail Processing Centre, South C...  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [12]:
df.shape

(103, 3)

The number of rows in this dataframe is 103.

## Section 2

In [13]:
pip install geocoder

Note: you may need to restart the kernel to use updated packages.


In [14]:
import geocoder

In [15]:
data = pd.read_csv("https://cocl.us/Geospatial_data")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

In [17]:
print("Wiki data shape: ", df.shape)
print("CSV data shape: ", data.shape)

Wiki data shape:  (103, 3)
CSV data shape:  (103, 3)


In [18]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [19]:
data.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [20]:
combined_data = df.join(data.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data

Postal Code           Borough  \
0           M3A        North York   
1           M4A        North York   
2           M5A  Downtown Toronto   
3           M6A        North York   
4           M7A  Downtown Toronto   
..          ...               ...   
98          M8X         Etobicoke   
99          M4Y  Downtown Toronto   
100         M7Y      East Toronto   
101         M8Y         Etobicoke   
102         M8Z         Etobicoke   

                                         Neighbourhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4          Queen's Park, Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100  Business reply mail Processing Centre, South C...  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

In [22]:
combined_data.shape

(103, 5)

103 rows for combined is a good result!

## Section 3

In [24]:
pip install geopy

Note: you may need to restart the kernel to use updated packages.


In [25]:
from geopy.geocoders import Nominatim

In [26]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


In [28]:
!pip install folium
import folium

Using folium to visualise the data

In [29]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for latitude, longitude, borough, neighbourhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

Looks good!

In [30]:
CLIENT_ID = 'CYHYRDROB33PYO2GPR5RQ444DIGXD1WH20N3WYM1KLZISSZT' 
CLIENT_SECRET = '2FVMDWMX2ZMK2DSLDOBVVRV5J2NBXJRFRIZ4DSXZRIM0GGS5'
VERSION = '20180605' # Foursquare API version

print('Credentials:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Credentials:
CLIENT_ID: CYHYRDROB33PYO2GPR5RQ444DIGXD1WH20N3WYM1KLZISSZT
CLIENT_SECRET:2FVMDWMX2ZMK2DSLDOBVVRV5J2NBXJRFRIZ4DSXZRIM0GGS5


All the venue categories:

In [31]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius
            )
            
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Category']
    
    return(nearby_venues)

Get nearby venues function for Toronoto neighbourhoods

In [32]:
venues_in_toronto = getNearbyVenues(combined_data['Neighbourhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [33]:
venues_in_toronto.shape

(1320, 5)

Five columns, and one thousand three hundred and twenty rows

Or, in numerics: Columns - 5; Rows - 1320

In [34]:
venues_in_toronto.head()

Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0         Parkwoods               43.753259               -79.329656   
1         Parkwoods               43.753259               -79.329656   
2  Victoria Village               43.725882               -79.315572   
3  Victoria Village               43.725882               -79.315572   
4  Victoria Village               43.725882               -79.315572   

                    Venue         Venue Category  
0         Brookbanks Park                   Park  
1           Variety Store      Food & Drink Shop  
2  Victoria Village Arena           Hockey Arena  
3               Portugril  Portuguese Restaurant  
4             Tim Hortons            Coffee Shop

In [35]:
venues_in_toronto.groupby('Neighbourhood').head()

Neighbourhood  \
0                                             Parkwoods   
1                                             Parkwoods   
2                                      Victoria Village   
3                                      Victoria Village   
4                                      Victoria Village   
...                                                 ...   
1305  Mimico NW, The Queensway West, South of Bloor,...   
1306  Mimico NW, The Queensway West, South of Bloor,...   
1307  Mimico NW, The Queensway West, South of Bloor,...   
1308  Mimico NW, The Queensway West, South of Bloor,...   
1309  Mimico NW, The Queensway West, South of Bloor,...   

      Neighbourhood Latitude  Neighbourhood Longitude                   Venue  \
0                  43.753259               -79.329656         Brookbanks Park   
1                  43.753259               -79.329656           Variety Store   
2                  43.725882               -79.315572  Victoria Village Arena   
3                  43.725882               -79.315572               Portugril   
4                  43.725882               -79.315572             Tim Hortons   
...                      ...                      ...                     ...   
1305               43.628841               -79.520999              Wingporium   
1306               43.628841               -79.520999        South St. Burger   
1307               43.628841               -79.520999               Dollarama   
1308               43.628841               -79.520999          Healthy Planet   
1309               43.628841               -79.520999    Artisano Bakery Café   

             Venue Category  
0                      Park  
1         Food & Drink Shop  
2              Hockey Arena  
3     Portuguese Restaurant  
4               Coffee Shop  
...                     ...  
1305            Wings Joint  
1306           Burger Joint  
1307         Discount Store  
1308        Supplement Shop  
1309                 Bakery  

[413 rows x 5 columns]

Looks good. Five columns and fourhundred thirteen rows
Columns - 5; Rows - 413

In [36]:
venues_in_toronto.groupby('Venue Category').max()

Neighbourhood  \
Venue Category                                                          
Accessories Store                    Lawrence Manor, Lawrence Heights   
Airport                                                     Downsview   
Airport Food Court  CN Tower, King and Spadina, Railway Lands, Har...   
Airport Gate        CN Tower, King and Spadina, Railway Lands, Har...   
Airport Lounge      CN Tower, King and Spadina, Railway Lands, Har...   
...                                                               ...   
Wine Bar                     Toronto Dominion Centre, Design Exchange   
Wine Shop                                Dufferin, Dovercourt Village   
Wings Joint         Mimico NW, The Queensway West, South of Bloor,...   
Women's Store                        Lawrence Manor, Lawrence Heights   
Yoga Studio                            University of Toronto, Harbord   

                    Neighbourhood Latitude  Neighbourhood Longitude  \
Venue Category                                                        
Accessories Store                43.718518               -79.464763   
Airport                          43.737473               -79.394420   
Airport Food Court               43.628947               -79.394420   
Airport Gate                     43.628947               -79.394420   
Airport Lounge                   43.628947               -79.394420   
...                                    ...                      ...   
Wine Bar                         43.653206               -79.379817   
Wine Shop                        43.669005               -79.442259   
Wings Joint                      43.628841               -79.520999   
Women's Store                    43.718518               -79.453512   
Yoga Studio                      43.715383               -79.321558   

                                              Venue  
Venue Category                                       
Accessories Store               Ardene Shoes Outlet  
Airport             Toronto Downsview Airport (YZD)  
Airport Food Court                Billy Bishop Café  
Airport Gate                                 Gate 8  
Airport Lounge                        Porter Lounge  
...                                             ...  
Wine Bar                          The National Club  
Wine Shop                                    Macedo  
Wings Joint                              Wingporium  
Women's Store                         Maximum Woman  
Yoga Studio                               YogaSpace  

[234 rows x 4 columns]

There are two hundred and thirty four rows, four columns.
Columns - 4; rows - 234

(Not sure if an Airport Gate is truly a venue, but I won't argue the point)

In [37]:
toronto_venue_cat = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")
toronto_venue_cat

Accessories Store  Airport  Airport Food Court  Airport Gate  \
0                     0        0                   0             0   
1                     0        0                   0             0   
2                     0        0                   0             0   
3                     0        0                   0             0   
4                     0        0                   0             0   
...                 ...      ...                 ...           ...   
1315                  0        0                   0             0   
1316                  0        0                   0             0   
1317                  0        0                   0             0   
1318                  0        0                   0             0   
1319                  0        0                   0             0   

      Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0                  0                0                 0                    0   
1                  0                0                 0                    0   
2                  0                0                 0                    0   
3                  0                0                 0                    0   
4                  0                0                 0                    0   
...              ...              ...               ...                  ...   
1315               0                0                 0                    0   
1316               0                0                 0                    0   
1317               0                0                 0                    0   
1318               0                0                 0                    0   
1319               0                0                 0                    0   

      Antique Shop  Aquarium  ...  Train Station  \
0                0         0  ...              0   
1                0         0  ...              0   
2                0         0  ...              0   
3                0         0  ...              0   
4                0         0  ...              0   
...            ...       ...  ...            ...   
1315             0         0  ...              0   
1316             0         0  ...              0   
1317             0         0  ...              0   
1318             0         0  ...              0   
1319             0         0  ...              0   

      Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                                 0                 0                      0   
1                                 0                 0                      0   
2                                 0                 0                      0   
3                                 0                 0                      0   
4                                 0                 0                      0   
...                             ...               ...                    ...   
1315                              0                 0                      0   
1316                              0                 0                      0   
1317                              0                 0                      0   
1318                              0                 0                      0   
1319                              0                 0                      0   

      Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                   0         0          0            0              0   
1                   0         0          0            0              0   
2                   0         0          0            0              0   
3                   0         0          0            0              0   
4                   0         0          0            0              0   
...               ...       ...        ...          ...            ...   
1315                0         0          0            0              0   
1316                0     

In [38]:
toronto_venue_cat['Neighbourhood'] = venues_in_toronto['Neighbourhood'] 

fixed_columns = [toronto_venue_cat.columns[-1]] + list(toronto_venue_cat.columns[:-1])
toronto_venue_cat = toronto_venue_cat[fixed_columns]

toronto_venue_cat.head()

Neighbourhood  Accessories Store  Airport  Airport Food Court  \
0         Parkwoods                  0        0                   0   
1         Parkwoods                  0        0                   0   
2  Victoria Village                  0        0                   0   
3  Victoria Village                  0        0                   0   
4  Victoria Village                  0        0                   0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  Antique Shop  ...  Train Station  \
0                    0             0  ...              0   
1                    0             0  ...              0   
2                    0             0  ...              0   
3                    0             0  ...              0   
4                    0             0  ...              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0                0         0          0            0              0   
1                0         0          0            0              0   
2                0         0          0            0              0   
3                0         0          0            0              0   
4                0         0          0            0              0   

   Yoga Studio  
0            0  
1            0  
2            0  
3            0  
4            0  

[5 rows x 235 columns]

Added neighbours to dataframed and shifted neighbourhood column

Now to group neighbourhoods and calculate mean

In [39]:
toronto_grouped = toronto_venue_cat.groupby('Neighbourhood').mean().reset_index()
toronto_grouped.head()

Neighbourhood  Accessories Store  \
0                                        Agincourt                0.0   
1                           Alderwood, Long Branch                0.0   
2  Bathurst Manor, Wilson Heights, Downsview North                0.0   
3                                  Bayview Village                0.0   
4                Bedford Park, Lawrence Manor East                0.0   

   Airport  Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0      0.0                 0.0           0.0             0.0              0.0   
1      0.0                 0.0           0.0             0.0              0.0   
2      0.0                 0.0           0.0             0.0              0.0   
3      0.0                 0.0           0.0             0.0              0.0   
4      0.0                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  ...  Train Station  \
0               0.0             0.000000           0.0  ...            0.0   
1               0.0             0.000000           0.0  ...            0.0   
2               0.0             0.000000           0.0  ...            0.0   
3               0.0             0.000000           0.0  ...            0.0   
4               0.0             0.043478           0.0  ...            0.0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                            0.0               0.0                    0.0   
1                            0.0               0.0                    0.0   
2                            0.0               0.0                    0.0   
3                            0.0               0.0                    0.0   
4                            0.0               0.0                    0.0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  \
0              0.0       0.0        0.0          0.0            0.0   
1              0.0       0.0        0.0          0.0            0.0   
2              0.0       0.0        0.0          0.0            0.0   
3              0.0       0.0        0.0          0.0            0.0   
4              0.0       0.0        0.0          0.0            0.0   

   Yoga Studio  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0  

[5 rows x 235 columns]

In [40]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
import numpy as np

Clustering the neighbourhood categories to get them down to a more reasonable number. Will use the top 10.

In [42]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighbourhood'] = toronto_grouped['Neighbourhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt                Lounge   
1                           Alderwood, Long Branch           Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North           Coffee Shop   
3                                  Bayview Village    Chinese Restaurant   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

       2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0  Latin American Restaurant        Clothing Store        Breakfast Spot   
1                Coffee Shop                   Gym          Skating Rink   
2                       Bank         Grocery Store           Pizza Place   
3                       Bank   Japanese Restaurant                  Café   
4             Sandwich Place    Italian Restaurant         Grocery Store   

  5th Most Common Venue    6th Most Common Venue        7th Most Common Venue  \
0      Department Store        Electronics Store  Eastern European Restaurant   
1                   Pub                    Diner                Deli / Bodega   
2                 Diner              Bridal Shop               Sandwich Place   
3           Yoga Studio         Department Store            Electronics Store   
4      Sushi Restaurant  Comfort Food Restaurant                     Pharmacy   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                    Drugstore            Donut Shop                Dog Run  
1             Department Store          Dessert Shop     Dim Sum Restaurant  
2                Deli / Bodega            Restaurant            Supermarket  
3  Eastern European Restaurant             Drugstore             Donut Shop  
4                  Pizza Place                  Café                    Pub

In [43]:
from sklearn.cluster import KMeans

In [44]:
k_num_clusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=k_num_clusters, random_state=0).fit(toronto_grouped_clustering)
kmeans

KMeans(n_clusters=5, random_state=0)

In [45]:
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Plotting preparation time

In [46]:
toronto_merged = combined_data

toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

toronto_merged.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656             1.0                  Park   
1  43.725882 -79.315572             0.0           Coffee Shop   
2  43.654260 -79.360636             0.0           Coffee Shop   
3  43.718518 -79.464763             0.0        Clothing Store   
4  43.662301 -79.389494             0.0           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0     Food & Drink Shop             Yoga Studio          Dance Studio   
1           Pizza Place   Portuguese Restaurant          Hockey Arena   
2                Bakery                    Park        Breakfast Spot   
3     Accessories Store  Furniture / Home Store              Boutique   
4      Sushi Restaurant             Yoga Studio        Sandwich Place   

         5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0  Eastern European Restaurant             Drugstore            Donut Shop   
1                  Curling Ice             Drugstore            Donut Shop   
2                      Theater           Yoga Studio                   Spa   
3                    Gift Shop           Event Space    Miscellaneous Shop   
4         Fast Food Restaurant    Italian Restaurant              Beer Bar   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0               Dog Run   Distribution Center         Discount Store  
1               Dog Run   Distribution Center         Discount Store  
2                  Café                   Pub         Farmers Market  
3           Coffee Shop         Women's Store  Vietnamese Restaurant  
4   Japanese Restaurant   Distribution Center         Discount Store

In [47]:
toronto_merged_nonan = toronto_merged.dropna(subset=['Cluster Labels'])

Removed NaN values

In [48]:
import matplotlib.cm as cm
import matplotlib.colors as colors

Set colour scheme, add markers, prepare for incoming map

In [49]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

x = np.arange(k_num_clusters)
ys = [i + x + (i*x)**2 for i in range(k_num_clusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged_nonan['Latitude'], toronto_merged_nonan['Longitude'], toronto_merged_nonan['Neighbourhood'], toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

Five clusters - let's check each cluster, one by one.

In [50]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 0, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1          North York             0.0           Coffee Shop   
2    Downtown Toronto             0.0           Coffee Shop   
3          North York             0.0        Clothing Store   
4    Downtown Toronto             0.0           Coffee Shop   
7          North York             0.0                   Gym   
..                ...             ...                   ...   
96   Downtown Toronto             0.0    Italian Restaurant   
97   Downtown Toronto             0.0                  Café   
99   Downtown Toronto             0.0           Coffee Shop   
100      East Toronto             0.0           Yoga Studio   
102         Etobicoke             0.0         Grocery Store   

    2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
1             Pizza Place   Portuguese Restaurant          Hockey Arena   
2                  Bakery                    Park        Breakfast Spot   
3       Accessories Store  Furniture / Home Store              Boutique   
4        Sushi Restaurant             Yoga Studio        Sandwich Place   
7              Beer Store              Restaurant   Japanese Restaurant   
..                    ...                     ...                   ...   
96             Restaurant                    Café           Coffee Shop   
97            Coffee Shop              Restaurant    Seafood Restaurant   
99       Sushi Restaurant                    Park                 Diner   
100           Pizza Place              Skate Park         Burrito Place   
102        Hardware Store             Flower Shop         Tanning Salon   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1             Curling Ice             Drugstore            Donut Shop   
2                 Theater           Yoga Studio                   Spa   
3               Gift Shop           Event Space    Miscellaneous Shop   
4    Fast Food Restaurant    Italian Restaurant              Beer Bar   
7             Coffee Shop    Chinese Restaurant                  Café   
..                    ...                   ...                   ...   
96                 Bakery                  Park            Playground   
97                  Hotel           Art Gallery        Sandwich Place   
99   Ethiopian Restaurant              Beer Bar   Japanese Restaurant   
100  Fast Food Restaurant                   Spa               Butcher   
102                Bakery         Burrito Place          Burger Joint   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                 Dog Run   Distribution Center         Discount Store  
2                    Café                   Pub         Farmers Market  
3             Coffee Shop         Women's Store  Vietnamese Restaurant  
4     Japanese Restaurant   Distribution Center         Discount Store  
7      Italian Restaurant   Sporting Goods Shop   Caribbean Restaurant  
..                    ...                   ...                    ...  
96             Beer Store        Sandwich Place                Butcher  
97                    Pub             Speakeasy                 Bakery  
99              Bookstore        Breakfast Spot        Bubble Tea Shop  
100            Restaurant                Garden          Auto Workshop  
102     Convenience Store  Fast Food Restaurant         Sandwich Place  

[85 rows x 12 columns]

First cluster good, now onto cluster two:

In [51]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 1, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0         North York             1.0                  Park   
21              York             1.0                  Park   
35         East York             1.0                  Park   
52        North York             1.0                  Park   
61   Central Toronto             1.0                  Park   
64              York             1.0                  Park   
66        North York             1.0                  Park   
83   Central Toronto             1.0                  Park   
85       Scarborough             1.0                  Park   
91  Downtown Toronto             1.0                  Park   
98         Etobicoke             1.0                  Park   

         2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Food & Drink Shop           Yoga Studio          Dance Studio   
21               Women's Store                  Pool          Dance Studio   
35           Convenience Store         Metro Station           Yoga Studio   
52                 Yoga Studio          Dance Studio     Electronics Store   
61                 Swim School              Bus Line           Yoga Studio   
64               Jewelry Store     Convenience Store           Yoga Studio   
66  Construction & Landscaping     Convenience Store           Yoga Studio   
83                       Trail                Lawyer           Yoga Studio   
85                  Playground          Intersection           Yoga Studio   
91                  Playground                 Trail           Yoga Studio   
98                       River           Yoga Studio          Dance Studio   

          5th Most Common Venue        6th Most Common Venue  \
0   Eastern European Restaurant                    Drugstore   
21  Eastern European Restaurant                    Drugstore   
35                Deli / Bodega  Eastern European Restaurant   
52  Eastern European Restaurant                    Drugstore   
61                Deli / Bodega  Eastern European Restaurant   
64                Deli / Bodega            Electronics Store   
66                Deli / Bodega            Electronics Store   
83                Deli / Bodega            Electronics Store   
85                 Dance Studio  Eastern European Restaurant   
91                  Curling Ice                    Drugstore   
98  Eastern European Restaurant                    Drugstore   

          7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0                    Donut Shop               Dog Run   Distribution Center   
21                   Donut Shop               Dog Run   Distribution Center   
35                    Drugstore            Donut Shop               Dog Run   
52                   Donut Shop               Dog Run   Distribution Center   
61                    Drugstore            Donut Shop               Dog Run   
64  Eastern European Restaurant             Drugstore            Donut Shop   
66  Eastern European Restaurant             Drugstore            Donut Shop   
83  Eastern European Restaurant             Drugstore            Donut Shop   
85                    Drugstore            Donut Shop               Dog Run   
91                   Donut Shop               Dog Run   Distribution Center   
98                   Donut Shop               Dog Run   Distribution Center   

   10th Most Common Venue  
0          Discount Store  
21         Discount Store  
35    Distribution Center  
52         Discount Store  
61    Distribution Center  
64                Dog Run  
66                Dog Run  
83                Dog Run  
85    Distribution Center  
91         Discount Store  
98         Discount Store

Cluster three:

In [52]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 2, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough             2.0  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
6         Deli / Bodega     Electronics Store  Eastern European Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6             Drugstore            Donut Shop               Dog Run   

  9th Most Common Venue 10th Most Common Venue  
6   Distribution Center         Discount Store

Fourth cluster

In [53]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 3, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             3.0        Baseball Field           Yoga Studio   
101   Etobicoke             3.0        Baseball Field           Yoga Studio   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57     Falafel Restaurant  Ethiopian Restaurant     Electronics Store   
101    Falafel Restaurant  Ethiopian Restaurant     Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th Most Common Venue 10th Most Common Venue  
57                Dog Run    Distribution Center  
101               Dog Run    Distribution Center

And lucky last, cluster 5:

In [54]:
toronto_merged_nonan.loc[toronto_merged_nonan['Cluster Labels'] == 4, toronto_merged_nonan.columns[[1] + list(range(5, toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
11  Etobicoke             4.0                Bakery           Yoga Studio   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
11    Falafel Restaurant  Ethiopian Restaurant     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
11  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
11               Dog Run    Distribution Center

Beautiful. Done!